R/DatabaseMigrations.R

```diff
@@ -0,0 +1,104 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+.migrationFileRexp <- "(Migration_[0-9]+)-(.+).sql"
+
+getExecutedMigrations <- function(connection, schema, tablePrefix) {
+  sql <- "
+    {DEFAULT @migration = migration}
+    SELECT migration_file FROM @result_schema.@table_prefix@migration ORDER BY migration_order;"
+  migrationsExecuted <- DatabaseConnector::renderTranslateQuerySql(connection,
+                                                                   sql = sql,
+                                                                   result_schema = schema,
+                                                                   table_prefix = tablePrefix,
+                                                                   snakeCaseToCamelCase = TRUE)
+
+  return(migrationsExecuted)
+}
+
+#' Migrate Data model
+#' @description
+#' Migrate data from current state to next state
+#'
+#' It is strongly advised that you have a backup of all data (either sqlite files, a backup database (in the case you
+#' are using a postgres backend) or have kept the csv/zip files from your data generation.
+#'
+#' @inheritParams createResultsDataModel
+#' @export
+migrateDataModel <- function(connection = NULL,
+                             connectionDetails = NULL,
+                             schema = NULL,
+                             tablePrefix = "") {
+  if (is.null(connection)) {
+    if (!is.null(connectionDetails)) {
+      connection <- DatabaseConnector::connect(connectionDetails)
+      on.exit(DatabaseConnector::disconnect(connection))
+    } else {
+      stop("No connection or connectionDetails provided.")
+    }
+  }
+
+  if (connection@dbms == "sqlite" & schema != "main") {
+    stop("Invalid schema for sqlite, use schema = 'main'")
+  }
+
+  migrationDir <- system.file("sql", "sql_server", "migrations", package = utils::packageName())
+  # List all migration files
+  allMigrations <- list.files(migrationDir, pattern = .migrationFileRexp)
+  # List files that have been executed
+  tables <- DatabaseConnector::getTableNames(connection, schema)
+  if (toupper(paste0(tablePrefix, "migration")) %in% tables) {
+    migrationsExecuted <- getExecutedMigrations(connection, schema, tablePrefix)
+    migrationsToExecute <- setdiff(allMigrations, migrationsExecuted$migrationFile)
+  } else {
+    migrationsToExecute <- allMigrations
+  }
+
+  ParallelLogger::logInfo("Executing database migrations on schema - ", schema)
+  migrationsToExecute <- sort(migrationsToExecute)
```

Comment:
Can you please confirm that `sort()` will handle the case where you have `Migration-1`, `Migration-11`, `Migration-2` and you get the result: `Migration-1`, `Migration-2`, `Migration-11` and not `Migration-1`, `Migration-11`, `Migration-2`.   If `sort()` is doing an alphabetical sort, then you'll get an out-of-order migration with 11 coming before 2.

Reply:
This is a great spot, thanks for the review!